Global Registry for Service Config Parsers

yashykt opened the pull request:
1) Creates a global registry for service config parsers.
2) Adds functions for resetting the registry (achieved by simply resetting the count of parsers).
3) Adds parsing logic for global and method params. (A lot of it is copying existing code. It is copy and not reuse, so that the old code can be simply deleted later).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/18586)
<!-- Reviewable:end -->

markdroth commented on src/core/ext/filters/client_channel/service_config.h:
```diff
@@ -96,6 +105,26 @@ class ServiceConfig : public RefCounted<ServiceConfig> {
+  /// registered parser. Each parser is responsible for reading the service
+  /// config json and returning a parsed object. This parsed object can later be
+  /// retrieved using the same index that was returned at registration time.
+  static int RegisterParser(ServiceConfigParser func) {
```
I like the idea of using the index as the key for the lookup!

markdroth commented on src/core/ext/filters/client_channel/service_config.h:
```diff
@@ -120,9 +149,16 @@ class ServiceConfig : public RefCounted<ServiceConfig> {
 
+  static constexpr int kMaxParsers = 32;
+  static int registered_parsers_count;
+  static ServiceConfigParser registered_parsers[kMaxParsers];
```
We'll need a way to initialize and clear this list when `grpc_init()` and `grpc_shutdown()` run.  I suggest using the same pattern that we use to register resolvers and LB policies.

yashykt replied:
done

markdroth commented on src/core/ext/filters/client_channel/service_config.h:
```diff
@@ -96,6 +105,26 @@ class ServiceConfig : public RefCounted<ServiceConfig> {
+    return parsed_service_config_objects[index].get();
+  }
+
+  typedef UniquePtr<ServiceConfigParsedObject> (*ServiceConfigParser)(
```
Instead of making the parser a simple function, I suggest making it a class.  It will need at least two methods, one for parsing the global parameters (replacing the mechanism currently provided by `ParseGlobalParams()`) and one for parsing per-method parameters (replacing the mechanism currently provided by `CreateMethodConfigTable()`).  Also, this will make it possible for the caller to pass any necessary external state to the parsing object via its ctor.

yashykt replied:
I was thinking of using the parser function to invoke the existing ParseGlobalParams and CreateMethodConfigTable, but making it a class will also work

markdroth commented on src/core/ext/filters/client_channel/service_config.h:
```diff
@@ -120,9 +149,16 @@ class ServiceConfig : public RefCounted<ServiceConfig> {
   UniquePtr<char> json_string_;  // Underlying storage for json_tree.
   grpc_json* json_tree_;
+
+  InlinedVector<UniquePtr<ServiceConfigParsedObject>, kMaxParsers>
```
I think we'll also need to store a method config table, where the value type of the table is this same type.

yashykt replied:
I was planning to let the method config table be a part of the service config parsed object

markdroth replied:
That would mean that each filter would have its own method config table, which means that each filter would have to independently look up the call's method in their own instance of the table.  It seems like it would be better for performance to have a single method config table whose value contains the per-method data for every filter.  That way, we can do the method config table lookup only once per call, in the client_channel filter.  We would then pass the resulting array of parsed objects into the subchannel stack, so each filter could grab its data via the index of its parser.

yashykt replied:
done!

yashykt commented:
Addressed the comments other than -
1) Adding tests - yet to add
2) Registering a UniquePtr for parsers to avoid copy. Given that the registration would be a one time thing compared to parsing which would happen more often, I don't see a reason to use a UniquePtr to avoid a simple copy which in this case should be something of a no-op since parsers would not have any state to them. Having a UniquePtr would just mean a layer of indirection everytime we want to parse service configs.

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 52 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-LbdxhFY6AWmT665lvvg:-LbePdPp73dACXFNLCmi:b-254rjp) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L52)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

If we make it required for the caller to pass in a non-null value for `error`, then I think we can just pass `error` in here directly, so we don't need `create_error` at all.  In fact, the entire rest of this method will go away -- we can just return the new service config directly (no need for `return_value`).
</blockquote></details>

I will do that when I update the rest of the code to use the error.


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 128 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe54a-49xo27rxnf3M:-LbeUaQ_DyoLXrbpSNuP:b45i4li) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L128)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Here's another place where we should ideally keep going after an error.
</blockquote></details>

If we don't get any names for this methodConfig, there is no more parsing to be done.


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 134 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe55Te7zVsJ7vQY49x:-LbeUjJM4ackUPtpnGPX:b45i4li) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L134)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Here's another place where we should ideally keep going after an error.
</blockquote></details>

If we don't get any names for this methodConfig, there is no more parsing to be done.


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 168 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe5HhAADmgeTzfN1ig:-LbeV8-F9ez3ZgBjtnSt:b-p432a2) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L168)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Here's another place where we should ideally keep going after an error.
</blockquote></details>

There is no more parsing to be done if methodConfig is not of array type


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 76 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe5h6s0D6EIk7HzxYQ:-LbeYFx0DByQZml2cTPb:b-lvs3ao) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L76)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This error message should probably be more specific.  If a user sees it, it needs to tell them what to fix in order to make the config work.

Same thing for all error messages here -- they should be specific about what the problem is and (as much as possible) where in the config the problem was.
</blockquote></details>

need to figure out  a good way to do that


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.h, line 216 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-LbdzMuSAJrVuDt7XmKK:-LbeZmoC4sd9yLr5URWY:b-l0gexe) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.h#L216)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

As per the previous comment, I think this separate storage vector should just go away.  But if it does stay, note that its expected size is actually not related to the number of parsers but rather the number of method parameter entries in the service config, so it should probably use something other than `kMaxParsers`.
</blockquote></details>

you're right, setting it to 32 for now


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.h, line 154 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-LbdvHxp-j2USU0rIUX-:-LbeZymi56VWFuP3jDLP:b-bhoq3b) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.h#L154)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

If we're not going to use a `UniquePtr<>` here, let's at least pass it in by value and then use `std::move()` when we call `push_back()`.  That way, we avoid the copy.
</blockquote></details>

Done. Though, the copy should not be different from a move here since the parser would not have any state. Passing it by value instead of a const ref would still have one copy.


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.h, line 89 at r3](https://reviewable.io/reviews/grpc/grpc/18586#-LbduqpR2XldFo4cvFkL:-Lbe_dtCDHVdMy2RpulB:b-7yk7dw) ([raw file](https://github.com/grpc/grpc/blob/0f0e774722f8d00682b6c2d6a05d4aebd22fd972/src/core/ext/filters/client_channel/service_config.h#L89)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think this is still needed.
</blockquote></details>

I was facing some strange issues earlier. Not facing them anymore. Adding it back


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 36 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbdww_61Yom3o9PgQoo:-LbiOmXv5e6kVY_bi684:b-vxf7po) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L36)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Can't this entire line go away?
</blockquote></details>

we need to define it somewhere since it is static.


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 36 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbdww_61Yom3o9PgQoo:-LbiOusG36Q35V54UVoU:b-m77k9k) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L36)):*
<details><summary><i>Previously, yashykt (Yash Tibrewal) wrote…</i></summary><blockquote>

we need to define it somewhere since it is static.
</blockquote></details>

there will be linker errors otherwise


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 134 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe55Te7zVsJ7vQY49x:-LbiPK_0EVXP_tTpsVFm:b8a8f73) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L134)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

That's true, but this failure doesn't mean that we didn't get any names; it just means that this particular name failed to parse.  We should still move on to the next name and try to parse it, so that we report all problems that exist with any name, not just the first one.
</blockquote></details>

You're right, done!


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 124 at r5](https://reviewable.io/reviews/grpc/grpc/18586#-LbhojBw54FozkNMnYuU:-LbiPMi0EOVR0e1BeFpi:bwr14wu) ([raw file](https://github.com/grpc/grpc/blob/390078b7a8a82b15de1c52a22e3e2d98b5843f24/src/core/ext/filters/client_channel/service_config.cc#L124)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This should probably use `grpc_error_add_child()`.
</blockquote></details>

Thanks for pointing it out


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 197 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbhpopd9n3C3VuxErEe:-LbiRGbc9Ued_i24KctM:bnejajh) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L197)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Was this line removed by mistake?
</blockquote></details>

the assert? earlier, it was possible for the two to be unequal. added it back


<!-- Sent from Reviewable.io -->

yashykt commented:
___
*[src/core/ext/filters/client_channel/service_config.cc, line 76 at r4](https://reviewable.io/reviews/grpc/grpc/18586#-Lbe5h6s0D6EIk7HzxYQ:-Lbivg_53m0l4hR-h6tJ:b-rlob0q) ([raw file](https://github.com/grpc/grpc/blob/c28f16e481993b27e996a7959a6c48385788976f/src/core/ext/filters/client_channel/service_config.cc#L76)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I realize that we don't have the kind of parser infrastructure to indicate what line number or column the error is on.  But we can at least provide scope information, like "in methodConfig entry for name {service=foo,method=bar}, field xxx failed parsing: reason".

Actually, instead of having to directly generate that kind of string, we can probably get a lot of this just from the error hierarchy.  For example, if the `ParsePerMethodParams()` method from a given parser returns an error saying "field xxx failed parsing: reason", then `ServiceConfig::ParsePerMethodParams()` can wrap that in an error that says "methodConfig entry for name {service=foo,method=bar} failed parsing", and the error hierarchy itself can help explain where the problem is.
</blockquote></details>

I like this!


<!-- Sent from Reviewable.io -->

yashykt commented:
Thanks for the detailed comments and review, Mark! They are very useful. I am going to continue other parts of the service config error handling implementation in subsequent PRs.
